third_party/txt/src/txt/paragraph_txt.cc

```diff
@@ -804,46 +838,54 @@ void ParagraphTxt::Layout(double width) {
+              ellipsis.length(), ellipsis.length(), run.is_rtl(), minikin_font,
+              minikin_paint, minikin_font_collection, nullptr);
+        // early_ellipsized is guaranteed that its line width exceed constraint.
+        // However, we still need to find the line_run that should do ellipsis.
+        //
+        // If current line_run can fit the constraint width and there is enough
+        // space for ellipsis, it is the job for next line_run to ellipsize.
+        // Including ellipsis_width will guarantee the next line_run can truncate
+        // enough space.
+        if (run_x_offset + text_width + ellipsis_width > width_ || line_run_it == line_runs.end() - 1) {
+          // Truncate characters from the text until the ellipsis fits.
+          size_t truncate_count = 0;
+          while (truncate_count < text_count &&
+                 run_x_offset + text_width + ellipsis_width > width_) {
+            text_width -= text_advances[text_count - truncate_count - 1];
+            truncate_count++;
+          }
 
-        // Truncate characters from the text until the ellipsis fits.
-        size_t truncate_count = 0;
-        while (truncate_count < text_count &&
-               run_x_offset + text_width + ellipsis_width > width_) {
-          text_width -= text_advances[text_count - truncate_count - 1];
-          truncate_count++;
-        }
-
-        ellipsized_text.reserve(text_count - truncate_count +
-                                ellipsis.length());
-        ellipsized_text.insert(ellipsized_text.begin(),
-                               text_.begin() + run.start(),
-                               text_.begin() + run.end() - truncate_count);
-        ellipsized_text.insert(ellipsized_text.end(), ellipsis.begin(),
-                               ellipsis.end());
-        text_ptr = ellipsized_text.data();
-        text_start = 0;
-        text_count = ellipsized_text.size();
-        text_size = text_count;
-
-        // If there is no line limit, then skip all lines after the ellipsized
-        // line.
-        if (paragraph_style_.unlimited_lines()) {
-          line_limit = line_number + 1;
-          did_exceed_max_lines_ = true;
+          ellipsized_text.reserve(text_count - truncate_count +
+                                  ellipsis.length());
+          ellipsized_text.insert(ellipsized_text.begin(),
+                                 text_.begin() + run.start(),
+                                 text_.begin() + run.start() + text_count - truncate_count);
+          ellipsized_text.insert(ellipsized_text.end(), ellipsis.begin(),
+                                 ellipsis.end());
+          text_ptr = ellipsized_text.data();
+          text_start = 0;
+          text_count = ellipsized_text.size();
+          text_size = text_count;
+
+          // If there is no line limit, then skip all lines after the ellipsized
+          // line.
+          if (paragraph_style_.unlimited_lines()) {
```

Comment:
If the line breaker is truncating all the line ranges beyond the ellipsis, then this logic will need to be updated.  I think this block can be removed, and the code that sets `did_exceed_max_lines_` at the start of `ParagraphTxt::Layout` should set it to true if the last line in `line_ranges_` has the ellipsis flag.

Reply:
yes, I forgot to clean up this part, thanks for catching